api/src/main/java/org/openmrs/module/htmlformentryui/element/ObsFromFragmentElement.java

```diff
@@ -0,0 +1,328 @@
+import org.openmrs.Obs;
+import org.openmrs.module.htmlformentry.FormEntryContext;
+import org.openmrs.module.htmlformentry.FormEntrySession;
+import org.openmrs.module.htmlformentry.FormSubmissionError;
+import org.openmrs.module.htmlformentry.HtmlFormEntryUtil;
+import org.openmrs.module.htmlformentry.action.FormSubmissionControllerAction;
+import org.openmrs.module.htmlformentry.element.HtmlGeneratorElement;
+import org.openmrs.module.htmlformentryui.tag.UiIncludeTagHandler;
+import org.openmrs.ui.framework.FragmentException;
+import org.openmrs.ui.framework.UiUtils;
+import org.openmrs.ui.framework.page.PageAction;
+
+public class ObsFromFragmentElement implements HtmlGeneratorElement, FormSubmissionControllerAction {
+
+	private UiUtils uiUtils;
+	private Map<String, Object> fragmentParams;
+	private Concept concept;
+	private String viewProvider;
+	private String fragment;
+	private String initFragmentParamName;
+	
+	// For tests only
+	public ObsFromFragmentElement() {
+		
+	}
+	
+	public ObsFromFragmentElement(Map<String, String> parameters, UiUtils uiUtils) {
+		this.uiUtils = uiUtils;
+		concept = HtmlFormEntryUtil.getConcept(parameters.get("conceptId"));
+		if (concept == null) {
+			throw new IllegalArgumentException("Cannot find concept with id " + parameters.get("conceptId"));
+		}
+		viewProvider = parameters.get("provider");
+		if (StringUtils.isBlank(viewProvider)) {
+			throw new IllegalArgumentException("Parameter provider cannot be blank");
+		}
+		fragment = parameters.get("fragment");
+		if (StringUtils.isBlank(fragment)) {
+			throw new IllegalArgumentException("Parameter fragment cannot be blank");
+		}
+		initFragmentParamName = parameters.get("initFragmentParamName");
+		if (StringUtils.isBlank(initFragmentParamName)) {
+			throw new IllegalArgumentException("Parameter initFragmentParamName cannot be blank");
+		}
+		StringBuilder fragmentUrlBuilder = new StringBuilder();
+		fragmentUrlBuilder.append(fragment);
+		fragmentUrlBuilder.append("?");
+		fragmentUrlBuilder.append(parameters.get("fragmentParams"));
+		try {
+			fragmentParams = UiIncludeTagHandler.paramsToMap(fragmentUrlBuilder.toString());
+							
+		} catch (URISyntaxException e) {
+			throw new FragmentException("Invalid fragment URI: " + fragmentUrlBuilder.toString() , e);
+		}
+	}
+	
+	@Override
+	public void handleSubmission(FormEntrySession session, HttpServletRequest request) {
+		String formFieldName = (String) fragmentParams.get("formFieldName");
+		if (formFieldName != null) {
+			String valueText = request.getParameter(formFieldName);
+			if (session.getContext().getMode().equals(FormEntryContext.Mode.EDIT)) {
+				List<Obs> existingObservations = session.getContext().removeExistingObs(concept);
+				if (existingObservations != null) {
+					if (!existingObservations.isEmpty()) {
+						// get the first Obs 
+						session.getSubmissionActions().modifyObs(existingObservations.get(0), concept, convertToType(valueText), new Date(), null);
```

Comment:
Can we clarify under which circumstance `existingObservations`'s size might be greater than 1, and hence what to do in such cases?

Reply:
This happens when someone uses one concept for more than one obs in the same Htmlformentry session. ie. 

``` 
	<obsFromFragment conceptId="165179" provider="uicommons" fragment="field/dropDown" fragmentParams="formFieldName=demo-field-label&amp;label=Label" initFragmentParamName="initialValue"/>
	<br/>
	<obsFromFragment conceptId="165179" provider="uicommons" fragment="field/radioButtons" fragmentParams="formFieldName=gender&amp;label=Gender" initFragmentParamName="initialValue"/>

```

When in the edit mode, `existingObservations.size() == 2` because concepts were reused; which I think is wrong!

Reply:
@samuelmale but this kind of stuff could happen:
```html
<obsgroup groupingConceptId="1">
  <obsFromFragment conceptId="165179" .../>
  ...
</obsgroup>

<obsgroup groupingConceptId="2">
  <obsFromFragment conceptId="165179" .../>
  ...
</obsgroup>
```

Reply:
Do you wanna educate me a little about the relationship btn the `groupingConcept` and the target `concept` for a particular obs.

I see you have put them in the `obsgroup` tag? I little bit perplexed!!!

Reply:
@samuelmale, first did you read the official docs about obs groups?

Reply:
yah sure. I understand that an `obsGroup` are logically grouped observations(https://wiki.openmrs.org/display/docs/Obs+Group) . 

Let's imagine we included two fragments using the `< obsFromFragment/>` tag but when they share the same concept; just like in the example you gave. 

While in the edit mode, we want to update the obs(possibly by creating a new obs?) the question here is; how would you tell right obs to update, since they share the same concept?

`List<Obs> existingObservations = session.getContext().removeExistingObs(concept);`

Yes the above method will return both the existing observations associated with this concept but how would you map the obsz to the fragments to update the right obs?

Reply:
I think I have a thought on this; I have strategy of how we can approach this :)

Anyway I would love to first hear your though or insight on this!